File: examples/src/imu_rotation_vector.cpp

#include <cstdio>
#include <iostream>

#include "utility.hpp"

// Inludes common necessary includes for development using depthai library
#include "depthai/depthai.hpp"

int main() {
    using namespace std;
    using namespace std::chrono;

    // Create pipeline
    dai::Pipeline pipeline;

    // Define sources and outputs
    auto imu = pipeline.create<dai::node::IMU>();
    auto xlinkOut = pipeline.create<dai::node::XLinkOut>();

    xlinkOut->setStreamName("imu");

    // Properties
    imu->enableIMUSensor({dai::IMUSensor::ROTATION_VECTOR}, 400);

    // Link plugins IMU -> XLINK
    imu->out.link(xlinkOut->input);

    dai::Device d(pipeline);

    auto imuQueue = d.getOutputQueue("imu", 50, false);
    auto baseTs = steady_clock::now();

    while(true) {
        auto imuData = imuQueue->get<dai::IMUData>();

        auto imuPackets = imuData->packets;
        for(auto& imuPacket : imuPackets) {
            auto rvTs = imuPacket.rotationVector.timestamp.get() - baseTs;
            auto& rVvalues = imuPacket.rotationVector;
            printf("Rotation vector timestamp: %ld ms\n", duration_cast<milliseconds>(rvTs).count());

            printf(
                "Quaternion: i: %.3f j: %.3f k: %.3f real: %.3f\n"
                "Accuracy (rad): %.3f \n",
                rVvalues.i,
                rVvalues.j,
                rVvalues.k,
                rVvalues.real,
                rVvalues.accuracy);
        }

        int key = cv::waitKey(1);
        if(key == 'q') {
            return 0;
        }
    }

    return 0;
}

File: examples/python/RVC2/IMU/imu_rotation_vector.py
#!/usr/bin/env python3
import depthai as dai

# Create pipeline
with dai.Pipeline() as pipeline:
    # Define sources and outputs
    imu = pipeline.create(dai.node.IMU)
    imu.enableIMUSensor(dai.IMUSensor.ROTATION_VECTOR, 50)
    # it's recommended to set both setBatchReportThreshold and setMaxBatchReports to 20 when integrating in a pipeline with a lot of input/output connections
    # above this threshold packets will be sent in batch of X, if the host is not blocked and USB bandwidth is available
    imu.setBatchReportThreshold(1)
    imu.setMaxBatchReports(10)

    imuQueue = imu.out.createOutputQueue(maxSize=50, blocking=False)

    pipeline.start()

    while pipeline.isRunning():
        imuData : dai.IMUData = imuQueue.get()
        imuPackets = imuData.packets
        for imuPacket in imuPackets:
            rotationVector = imuPacket.rotationVector
            imuF = "{:.06f}"
            print(f"Rotation Vector: i: {imuF.format(rotationVector.i)} j: {imuF.format(rotationVector.j)} k: {imuF.format(rotationVector.k)} real: {imuF.format(rotationVector.real)}")